pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,282 @@
+		minDistance := float64(len(requirements) + 1)
+		minPod := 0
+		for i, pod := range pods {
+			dist := requirements.distance(pod)
+			if dist < minDistance || minDistance == 0 && dist == 0 && smallerResourceRequest(pod, pods[minPod]) {
+				minDistance = dist
+				minPod = i
+			}
+		}
+		// subtract the pod from requirements, and transfer the pod from input-pods to pods-to-evicted
+		requirements = requirements.subtract(pods[minPod])
+		podsToEvict = append(podsToEvict, pods[minPod])
+		pods[minPod] = pods[len(pods)-1]
+		pods = pods[:len(pods)-1]
+	}
+	return podsToEvict, nil
+}
+
+type admissionRequirement struct {
+	resourceName v1.ResourceName
+	quantity     int64
+}
+
+type admissionRequirementList []*admissionRequirement
+
+// distance of the pods requests from the admissionRequirements.
+// distance is measured by the fraction of the requirement satisfied by the pod,
+// so that each requirement is weighted equally, regardless of absolute magnitude.
+func (a admissionRequirementList) distance(pod *v1.Pod) float64 {
+	dist := float64(0)
+	for _, req := range a {
+		remainingRequest := float64(req.quantity - getResourceRequest(pod, req.resourceName))
+		if remainingRequest < 0 {
+			remainingRequest = 0
+		}
+		dist += math.Pow(remainingRequest/float64(req.quantity), 2)
+	}
+	return dist
+}
+
+// returns a new admissionRequirementList containing remaining requirements if the provided pod
+// were to be preempted
+func (a admissionRequirementList) subtract(pods ...*v1.Pod) admissionRequirementList {
+	newList := []*admissionRequirement{}
+	for _, req := range a {
+		newQuantity := req.quantity
+		for _, pod := range pods {
+			newQuantity -= getResourceRequest(pod, req.resourceName)
+		}
+		if newQuantity > 0 {
+			newList = append(newList, &admissionRequirement{
+				resourceName: req.resourceName,
+				quantity:     newQuantity,
+			})
+		}
+	}
+	return newList
+}
+
+func (a admissionRequirementList) toString() string {
+	s := "["
+	for _, req := range a {
+		s += fmt.Sprintf("(res: %v, q: %d), ", req.resourceName, req.quantity)
+	}
+	return s + "]"
+}
+
+// returns pods in the list of pods where filterFunc returns true.
+func filterPods(pods []*v1.Pod, filterFunc func(p *v1.Pod) bool) []*v1.Pod {
+	filteredPods := []*v1.Pod{}
+	for _, pod := range pods {
+		if filterFunc(pod) {
+			filteredPods = append(filteredPods, pod)
+		}
+	}
+	return filteredPods
+}
+
+// returns true if pod1 has a smaller request than pod2
+func smallerResourceRequest(pod1 *v1.Pod, pod2 *v1.Pod) bool {
+	priorityList := []v1.ResourceName{
+		v1.ResourceNvidiaGPU,
+		v1.ResourceMemory,
+		v1.ResourceCPU,
+	}
+	for _, res := range priorityList {
+		req1 := getResourceRequest(pod1, res)
+		req2 := getResourceRequest(pod2, res)
+		if req1 < req2 {
+			return true
+		} else if req1 > req2 {
+			return false
+		}
+	}
+	return true
+}
+
+// finds and returns the request for a specific resource.
+func getResourceRequest(pod *v1.Pod, resource v1.ResourceName) int64 {
```

Comment:
This function is useful and I suspect the scheduler needs something similar. Let's try to share this method by placing it in some common lib.